### PR TITLE
feat: introduce Arrow-based function chain pipeline to replace legacy rerank implementation

### PR DESCRIPTION
**Pull request by junjiejiangjjj**
https://github.com/milvus-io/milvus/issues/46565
design doc: https://github.com/milvus-io/milvus-design-docs/blob/main/design_docs/20260310-function-chain.md

Adds a new chain package (internal/util/function/chain/) that implements a composable, Apache Arrow-based pipeline for search result post-processing (reranking, scoring, merging, filtering, grouping, etc.). This replaces the legacy rerank utility functions with a DataFrame-oriented approach that provides better performance and extensibility.

Key changes:
  - New chain package with DataFrame abstraction, operator registry, expression engine (decay, score combine, round decimal, rerank model), and operators (merge, sort, filter, limit, select, map, group-by)
  - Refactored search_pipeline.go to use chain-based reranking instead of the legacy rerank package, adding a hybrid_assemble operator for hybrid search
  - Removed legacy rerank utilities (decay_function, rrf_function, weighted_function, util.go, rerank_base) and simplified remaining rerank providers to thin wrappers
  - Added rerank_meta.go in proxy for structured rerank configuration parsing
  - Comprehensive unit tests and benchmarks for the chain package
  - Extended Python and Go integration tests for reranker functions

### CODE REVIEW

**Review by liliu-z — COMMENTED** [review]
## Code Review Summary

**Overall**: 架构方向正确（Arrow 列式 + 可组合算子链），核心数学公式与旧代码字符级一致，Arrow 内存管理系统性做对。但存在若干需修复的问题。

### 🔴 必须修复（6 项）

**S1. Model reranker + group_by 校验被删除** (`task_search.go`)
旧代码的 `IsSupportGroup()` 检查被完全删除且无替代。Model reranker + group_by 组合会静默产生语义错误的结果。

**S2. `normalize` 参数被静默丢弃** (`rerank_builder.go`)
`parseDecayParams` 返回的 `normalize` 被 `_` 忽略，`buildDecayChain` 硬编码 `WithNormalize(true)`。用户设置 `norm_score=false` 无效且无提示。

**S3. `NewLimitOp` 负数参数导致 panic** (`operator_limit.go`)
直接构造函数不验证负数，负 limit 导致 `array.NewSlice(data, -1, -2)` panic。

**S4. 参数校验从 PreExecute 推迟到 PostExecute** (`task_search.go`)
旧代码在 `initAdvancedSearchRequest` 阶段校验 rerank 参数，新代码推迟到运行时。无效参数不再被提前拒绝。

**S5. 缺少设计文档**（已标记 `do-not-merge/missing-design-doc`）

**S6. `NewGroupByOpFromRepr` 负 offset 导致 slice panic** (`operator_group_by.go`)
Fluent API 有保护，但 `FromRepr` 路径缺少 `offset < 0` 校验，负 offset 导致 `groups[negativeIndex:]` panic。

### 🟡 建议修复（11 项）

- **M1**: `compareArrayValues` 不支持 `LARGE_STRING` 但 `isComparableType` 声明支持 → 排序不生效 (`operator_sort.go`)
- **M2**: 全链路无 NaN/Inf 防护，从 DecayExpr 静默传播到最终结果 (`decay_expr.go` 等)
- **M3**: `GetInputFieldNamesFromFuncScore` 固定取 `Functions[0]`，boost 排首位时取错 (`rerank_builder.go`)
- **M4**: null group key 被归入同一组（`nil` 作为 `map[any]` key）(`operator_group_by.go`)
- **M5**: `DataFrame.Release()` 后 `nameIndex` 残留，误用可 panic (`dataframe.go`)
- **M6**: null 文本当空字符串发给外部 rerank 模型 (`rerank_model_expr.go`)
- **M7**: `convertLegacyParams` 使用 `reflect` 做类型检查，直接 `v.(float64)` 即可 (`rerank_builder.go`)
- **M8**: `search_reduce_util.go` GroupBy 解包遗漏 `IsDynamic` 等元数据字段
- **M9**: `hybridAssembleOp` 按位置索引追加字段数据，子搜索字段布局不一致时错位 (`search_pipeline.go`)
- **M10**: MergeOp 未校验同名列跨 DataFrame 的数据类型一致性，类型不匹配时静默返回零值 (`operator_merge.go`)
- **M11**: `orderBy + rerank` 组合时 rerank 被静默跳过，无冲突校验 (`search_pipeline.go`)

### 🔵 测试覆盖缺口（8 项）

- **T1**: 没有旧/新实现输出一致性对比测试 — 26K 行重构最关键的安全网
- **T2**: 没有并发安全测试
- **T3**: Benchmark 没有新旧性能对比（PR 声称性能更好但无数据）
- **T4**: 端到端缺少精确 ID 排序验证（旧代码 `TestRerankProcess` 有，迁移后丢失）
- **T5**: Boost ranker Python 集成测试断言过弱（大多数只检查"搜索成功"）
- **T6**: `search_pipeline_test.go` 降低了覆盖度（roundDecimal 改 -1、fillFieldNames 路径未覆盖）
- **T7**: 含空 DataFrame 的多输入合并场景未测试
- **T8**: 可空稀疏向量索引映射测试未迁移

### 🟢 代码质量（8 项）

- `normalizeScoreChunk` 死代码（定义但未被调用，仅测试引用）
- `types.AllStages` 是可变全局 `var`
- `validateOperators` 硬编码类型断言（建议接口加 `Validate` 方法）
- `MergeOp` 不可序列化，打破框架统一性（已有 TODO）
- `parseDecayParams` 和 `buildWeightedChain` 权重验证重复
- `plan_parser_v2_test.go` 硬编码字符串替代常量
- `getArrayValue` 定义在 merge 中但仅被 group_by 调用
- RRF 中 `k` 参数精度从 float32 升级为 float64（微小行为差异）

### ✅ 做得好的地方

- Arrow 引用计数管理系统性正确，所有退出路径无泄漏，`CheckedAllocator` 测试充分
- 核心数学公式（RRF、4 种归一化、加权合并）与旧代码字符级一致
- 修复了旧代码中 MHJACCARD/MaxSim 等度量被错误当作距离度量处理的 bug
- 批量分片迁移增加了 `maxBatch <= 0` 防死循环保护
- 关注点分离清晰（Expr vs Operator vs Chain），16K 行高质量测试

**Comment by junjiejiangjjj** [comment]
## 🟢 Code Quality

**1. `normalizeScoreChunk` dead code** ✅ Fixed

Confirmed it was only referenced by tests. Removed the function and its tests.

**2. `types.AllStages` is a mutable global `var`**

`AllStages` is only referenced during expr construction (not on the hot path), and no code mutates it. Converting it to a function that returns a copy would add unnecessary allocations. Keeping it as a `var` is fine.

**3. `validateOperators` hardcoded type assertions**

Agreed this could be cleaner with a `Validate` interface method. However, there are only 3 type assertions currently, so the scope is manageable. Will consider refactoring when adding new operator types.

**4. `MergeOp` not serializable**

Acknowledged — there is already a TODO in the code explaining why: MergeOp requires runtime context (metricTypes, weights, rrfK, normalize) that cannot be recovered from a static OperatorRepr. Will address when the serialization framework evolves.

**5. Weight validation duplicated between `parseWeightedParams` and `buildWeightedChain`** ✅ Fixed

Removed the redundant weight range check from `buildWeightedChain`. `parseWeightedParams` is the single validation point now.

**6. `plan_parser_v2_test.go` hardcoded strings instead of constants**

Only one instance found (`"weighted"` at line 1928). Too minor to warrant a change.

**7. `getArrayValue` defined in merge but only called by group_by** ✅ Fixed

Moved `getArrayValue` to `operator_base.go` alongside other shared Arrow helpers like `sliceArray`.

**8. RRF `k` parameter precision upgraded from float32 to float64**

This is intentional. Float64 provides better intermediate precision. The final result is still cast back to float32 (`float32(1.0 / (op.rrfK + float64(...)))`), so output type is consistent with the old implementation. This is an improvement, not a regression.

**Review by liliu-z — COMMENTED** [review]
Code review: 7 issues identified (2 critical regressions, 5 important improvements)

File: internal/util/function/chain/rerank_builder.go [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+type SearchParams struct {
+	Nq           int64
+	Limit        int64
+	Offset       int64
+	RoundDecimal int64
+
+	// Grouping parameters
+	GroupByField string      // Field to group by (empty means no grouping)
+	GroupSize    int64       // Maximum rows per group
+	GroupScorer  GroupScorer // How to compute group score ("max", "sum", "avg")
```
Comment (liliu-z) [inline internal/util/function/chain/rerank_builder.go:88]: 🔴 **`strictGroupSize` missing from SearchParams — API regression**

`strictGroupSize` is parsed in proxy (`search_util.go:33`) and stored in `rankParams`, but `rerankOperator` never reads it, `SearchParams` doesn't define it, and `GroupByOp` doesn't enforce it. Users setting `strictGroupSize=true` will silently get non-strict results after this change.

File: internal/util/function/chain/converter.go [hunk 2]
```diff
@@ -0,0 +1,834 @@
...
+		data, err := getScalarStringData(fieldData, fieldName)
+		if err != nil {
+			return err
+		}
+		if err := validateLen(len(data)); err != nil {
+			return err
+		}
+		chunks = importChunkedBatch(data, offsets, getValidSlice, array.NewStringBuilder, alloc)
+
+	default:
```
Comment (liliu-z) [inline internal/util/function/chain/converter.go:509]: 🔴 **`Timestamptz` type not handled — regression**

The old `rerank/util.go` handled `DataType_Timestamptz` as `int64`. This type switch covers Bool through String/VarChar/Text but has no Timestamptz case. Collections using Timestamptz fields for decay rerank will hit this `default` error branch. Fix: add `case schemapb.DataType_Timestamptz:` handled like Int64.

File: internal/util/function/chain/expr/score_combine_expr.go [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+		hasNull := false
+		for _, input := range inputs {
+			if input.Chunk(chunkIdx).IsNull(rowIdx) {
+				hasNull = true
+				break
+			}
+		}
+
+		if hasNull {
+			builder.AppendNull()
```
Comment (liliu-z) [inline internal/util/function/chain/expr/score_combine_expr.go:206]: 🟡 **NaN silently propagated to user output**

Null inputs are correctly handled here (AppendNull), but NaN floats from upstream pass through all combine modes unchecked. Built-in decay expressions have parameter validation that prevents NaN production, but there's no defensive layer if a future expression produces NaN. Consider adding `math.IsNaN(result)` check after `s.combine(values)` and returning an error or substituting 0.

File: internal/util/function/chain/operator_group_by.go [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+	key        any       // Group key value
+	rowIndices []int     // Row indices belonging to this group
+	groupScore float32   // Computed group score
+	scores     []float32 // Individual scores for sum/avg computation
+	ids        []any     // ID values for tiebreaking
+}
+
+// buildGroups builds groups from the chunk.
+func (o *GroupByOp) buildGroups(groupChunk arrow.Array, scoreChunk *array.Float32, idChunk arrow.Array, chunkLen int) []*group {
+	groupMap := make(map[any]*group)
```
Comment (liliu-z) [inline internal/util/function/chain/operator_group_by.go:347]: 🟡 **No group cardinality bound**

This map grows to match the number of distinct group keys with no upper limit. While TopK (≤16384) caps it in practice, grouping by a near-unique field silently produces degenerate results (one row per group). A `log.Warn` when `len(groupMap) > limit * 10` would help users catch misconfigured group-by fields early.

File: internal/util/function/chain/rerank_builder.go [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+		if err := buildModelChain(fc, collSchema, funcSchema, searchMetrics, searchParams); err != nil {
+			return nil, err
+		}
+
+	default:
+		return nil, merr.WrapErrParameterInvalidMsg("rerank_builder: unsupported reranker %s", rerankerName)
+	}
+
+	// Add grouping or sort+limit
+	if searchParams.HasGrouping() {
```
Comment (liliu-z) [inline internal/util/function/chain/rerank_builder.go:314]: 🟡 **`IsSupportGroup` check removed — intentional?**

The old code rejected group-by for incompatible rerankers via `IsSupportGroup()` in `task_search.go`. The new architecture decouples GroupByOp from rerankers, implicitly making all rerankers group-compatible. If intentional, a comment here explaining the design change would help — otherwise this is a missing validation.

File: internal/util/function/chain/operator_merge.go [hunk 6]
```diff
@@ -0,0 +1,968 @@
...
+}
+
+// mergeRRF implements Reciprocal Rank Fusion.
+func (op *MergeOp) mergeRRF(ctx *types.FuncContext, inputs []*DataFrame) (*DataFrame, error) {
+	return op.mergeWithScoreCollector(ctx, inputs, op.collectRRFScores)
+}
+
+// collectRRFScores collects RRF scores for a single chunk.
+func (op *MergeOp) collectRRFScores(inputs []*DataFrame, chunkIdx int) (map[any]float32, map[any]idLocation, error) {
+	idScores := make(map[any]float32)
```
Comment (liliu-z) [inline internal/util/function/chain/operator_merge.go:321]: 🟡 **`map[any]` loses compile-time type safety**

`map[any]float32` relies on runtime type consistency of ID keys. If different chunks are ever parsed into mismatched types (e.g., `int64` vs `string`), lookups silently fail. A type-switched dispatch to `map[int64]` / `map[string]` would catch this at compile time.

File: internal/util/function/chain/operator_sort.go [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+		return compareTyped(a, i, j)
+	case *array.Uint64:
+		return compareTyped(a, i, j)
+	case *array.Float32:
+		return compareTyped(a, i, j)
+	case *array.Float64:
+		return compareTyped(a, i, j)
+	case *array.String:
+		return compareTyped(a, i, j)
+	default:
```
Comment (liliu-z) [inline internal/util/function/chain/operator_sort.go:252]: 🟡 **`default: return 0` hides future bugs**

If a new type is added to `isComparableType()` but not to `compareArrayValues()`, rows with that type silently compare as equal, producing wrong sort order with no error. Since reaching this branch indicates an internal logic inconsistency, this should be a panic or error rather than a silent fallback.

File: internal/util/function/chain/rerank_builder.go (hunk 1 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/rerank_builder.go:88]: `strictGroupSize` is a QueryNode-level parameter that controls segment-level search behavior — whether each group must be fully filled to `groupSize` rows before the search considers it "enough". The proxy only parses it and passes it through to QueryNode via the `QueryInfo` proto. The old rerank code never used it on the proxy side either. The chain code operates at the proxy rerank/merge layer and does not need to handle this parameter.

File: internal/util/function/chain/converter.go (hunk 2 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/converter.go:509]: `Timestamptz` is a newly added data type. The scope of this PR is to cover all existing capabilities of the old rerank implementation. Support for new types like `Timestamptz` will be added in a follow-up.

File: internal/util/function/chain/expr/score_combine_expr.go (hunk 3 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/expr/score_combine_expr.go:206]: The decay parameter validation in `NewDecayExpr` already prevents NaN production: `decay` is constrained to (0.001, 0.999) so `log(decay)` is always a finite negative, `scale > 0` prevents division by zero, and `offset >= 0`. All three decay functions are mathematically bounded to (0, 1] under these constraints. The only theoretical path to NaN would be a user storing NaN in a float field — the old rerank code had no guard for that either. Acknowledged as a defensive improvement, but not a real risk in practice.

File: internal/util/function/chain/operator_group_by.go (hunk 4 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/operator_group_by.go:347]: The input size is already bounded by upstream TopK (≤16384), so the map cannot grow unbounded. The group-by field is user-specified — if they pick a near-unique field, the degenerate result is expected behavior, same as the old code. A warning log could be useful for observability but is not a correctness issue. Will consider adding it as a follow-up improvement.

File: internal/util/function/chain/rerank_builder.go (hunk 5 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/rerank_builder.go:314]: Yes, this is intentional. The new architecture decouples group-by into a standalone `GroupByOp` that works uniformly across all rerankers, so individual rerankers no longer need to declare group-by compatibility via `IsSupportGroup()`. This was one of the design goals of the refactor — eliminate per-reranker group-by logic in favor of a single composable operator.

File: internal/util/function/chain/operator_merge.go (hunk 6 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/operator_merge.go:321]: In Milvus, the ID (primary key) type is defined at the collection level — either Int64 or VarChar. All inputs to MergeOp come from sub-searches on the same collection, so the ID type is guaranteed to be consistent across all inputs. The type mismatch scenario (int64 vs string keys in the same map) cannot happen in practice.

File: internal/util/function/chain/rerank_builder.go [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+	}
+
+	return weights, normalize, nil
+}
+
+// =============================================================================
+// Decay Builder
+// =============================================================================
+
+func buildDecayChain(fc *FuncChain, collSchema *schemapb.CollectionSchema, funcSchema *schemapb.FunctionSchema, searchMetrics []string) error {
```
Comment (liliu-z) [inline internal/util/function/chain/rerank_builder.go:458]: buildDecayChain uses fc.Merge which loses sortDescending propagation. For L2 + norm_score=false, results are incorrectly sorted DESC instead of ASC.

File: internal/util/function/chain/rerank_builder.go [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+	}
+
+	// Add grouping or sort+limit
+	if searchParams.HasGrouping() {
+		// Use GroupBy for grouping search
+		scorer := searchParams.GroupScorer
+		if scorer == "" {
+			scorer = GroupScorerMax
+		}
+		fc.GroupByWithScorer(searchParams.GroupByField, searchParams.GroupSize, searchParams.Limit, searchParams.Offset, scorer)
```
Comment (liliu-z) [inline internal/util/function/chain/rerank_builder.go]: GroupByWithScorer path does not consume sortDescending and hard-codes DESC, breaking weighted + group_by + L2 + normalize=false.

File: internal/proxy/search_pipeline.go [hunk 10]
```diff
@@ -578,6 +758,86 @@ func pickFieldData(ids *schemapb.IDs, pkOffset map[any]int, fields []*schemapb.F
...
+
+	// Pre-compute field-index computers per sub-search result (one per distinct FieldsData layout).
+	computers := make([]*typeutil.FieldDataIdxComputer, len(reducedResults))
+	for i, r := range reducedResults {
+		if len(r.GetResults().GetFieldsData()) > 0 {
+			computers[i] = typeutil.NewFieldDataIdxComputer(r.GetResults().GetFieldsData())
+		}
+	}
+
+	// Assemble only the rows referenced by the reranked IDs.
```
Comment (liliu-z) [inline internal/proxy/search_pipeline.go:823]: Skipping nil computers without removing corresponding IDs/FieldsData entries causes row/column misalignment in assembled results.

File: internal/util/function/chain/rerank_builder.go [hunk 11]
```diff
@@ -0,0 +1,718 @@
...
+
+// GetRerankNameFromFuncScore returns the reranker name from a FunctionScore schema.
+func GetRerankNameFromFuncScore(funcScore *schemapb.FunctionScore) string {
+	if funcScore == nil || len(funcScore.Functions) == 0 {
+		return ""
+	}
+	return rerank.GetRerankName(funcScore.Functions[0])
+}
+
+func validateInputField(collSchema *schemapb.CollectionSchema, fieldName string) error {
```
Comment (liliu-z) [inline internal/util/function/chain/rerank_builder.go:724]: validateInputField accepts Timestamptz but GetNumericValue only handles Int/Float, causing runtime failures after passing schema validation.

File: internal/util/function/chain/rerank_builder.go (hunk 8 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/rerank_builder.go:458]: Confirmed and fixed. Root cause is slightly different from "lost sortDescending propagation": decay multiplies `$score` by a `[0, 1]` factor and assumes "higher = better", so the chain must always produce DESC scores. Legacy `rerank/decay_function.go` enforced this by passing `toGreater=true` to `getNormalizeFunc` regardless of `norm_score`; the chain refactor lost that.

Fix: added `WithForceDescending` option on `MergeOp`. When set, `resolveMergeBehavior` applies `getDirectionConvertFunc` (atan-based) for distance metrics regardless of `normalize`. `buildDecayChain` always passes `WithForceDescending(true)`. Other rerankers unchanged.

Coverage: `TestExecuteDecay_L2_NoNormScore_RanksByCombinedScore` (asserts each score matches `(1 − 2·atan(d)/π) × gauss_decay` within `1e-5`) + Python e2e `test_milvus_client_search_decay_rerank_l2_metric_no_norm_score`.

File: internal/util/function/chain/rerank_builder.go (hunk 9 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/rerank_builder.go]: Confirmed. This is also a long-standing latent bug in legacy `rerank/util.go::newGroupingIDScores` (no direction parameter, hard-coded DESC) — the chain refactor inherited it. Fix:

1. Added `sortDescending bool` field on `GroupByOp` (defaults `true`, preserving legacy behavior for all existing callers and `NewGroupByOpFromRepr`)
2. Added builder-style `SetSortDescending(bool)` setter
3. Direction-aware `sortAndLimitGroup` (within-group trim) and `processChunk` step 4 (cross-group sort)
4. `buildRerankChainInternal` constructs the op directly and propagates `sortDescending` from `buildWeightedChain`

`Max` scorer code is unchanged: `scores[0]` is now correctly the best representative under either direction since `sortAndLimitGroup` orders the slice first. Public chain API (`GroupBy`/`GroupByWithScorer`) unchanged — no breaking change.

While validating this end-to-end I also discovered a related pre-existing proxy bug in `reduceAdvanceGroupBy` single-shard early return path (`search_reduce_util.go:86`) — without that fix the chain layer fix alone wasn't observable from Python e2e because the proxy was passing `-L2` to the chain. Also fixed in this PR with its own table-driven test (TDD-validated by reverting the fix and watching the new test go red, then re-applying).

Coverage: `TestExecuteWeightedGroupBy_L2_NoNormScore_PreservesAscOrder`, `TestExecuteWeightedGroupBy_L2_NormScore_PreservesDescOrder`, `TestGroupByOp_AscDirection`, `TestGroupByOp_DefaultSortDescending` + Python e2e `test_milvus_client_hybrid_search_weighted_groupby_l2_no_norm_score` + new `TestReduceAdvanceGroupBy_SingleShardScoreNegation` and `TestReduceAdvanceGroupBy_SingleShardMatchesMultiShard`.

File: internal/proxy/search_pipeline.go (hunk 10 above)
Comment (junjiejiangjjj) [reply internal/proxy/search_pipeline.go:823]: You're right that this is an invariant violation. Tracing the proxy code, the precondition (one sub-result with `FieldsData` while another has none) shouldn't be reachable today — `task_search.go` sets identical `plan.OutputFieldIds` across all sub-requests and the `needRequery=false` path used by `hybridSearchPipe` always includes the PK. So the defensive `continue` is dead code that would only fire on an upstream invariant violation, where silently dropping the row is the worst possible outcome.

Fix: `continue` → `return merr.WrapErrServiceInternal(...)` with the offending sub-result index, the PK that triggered it, and the collection ID. Surfacing the bug at its source is much better than corrupting `Ids ↔ FieldsData` row alignment.

Coverage: `TestHybridAssembleOp_MixedFieldsDataLayoutErrors` (synthesizes the offending state and asserts the operator returns an error mentioning `FieldsData`).

File: internal/util/function/chain/rerank_builder.go (hunk 11 above)
Comment (junjiejiangjjj) [reply internal/util/function/chain/rerank_builder.go:724]: Investigated: legacy decay listed `Timestamptz` in its type-dispatch switch (added in `c0d62268a`) but the support never reached end-to-end. No production path or test ever exercised it — `getField` had a branch but neither converter nor `GetNumericValue` handled it, and there was no decay+Timestamptz test. The chain refactor reproduces the same incomplete state.

Since this PR's scope is to preserve legacy functionality rather than extend it, I removed `Timestamptz` from `validateInputField` so the chain rejects it explicitly. The user-visible error in proxy e2e actually comes from `chain.FromSearchResultData` (Arrow converter `default` branch) which runs before `BuildRerankChain` — both layers reject `Timestamptz`, just at different stages. Adding genuine `Timestamptz` support (converter + `GetNumericValue` + tests) belongs in a separate feature PR.

Coverage: `TestBuildDecayChainTimestamptzInputField` (chain unit) + Python e2e `test_milvus_client_search_decay_rerank_timestamptz_field_rejected` (asserts `unsupported field type: Timestamptz`).

**Comment by liliu-z** [comment]
**internal/proxy/search_pipeline.go:1** strictGroupSize is silently dropped in the Arrow chain path, causing a functional regression between single-path and multi-path search. Must be restored before merge.

**Comment by liliu-z** [comment]
**internal/proxy/rerank/converter.go:280** When ids==nil && totalRows>0, a malformed DataFrame is silently constructed instead of returning an error.

**Comment by liliu-z** [comment]
**internal/proxy/rerank/converter.go:293** Same hole as ids branch: when totalRows>0 && len(scores)==0, the $score column is silently dropped. Return an explicit error.

**Comment by liliu-z** [comment]
**internal/proxy/rerank/converter.go:1** Legacy rerank/ directory is nearly emptied but not fully removed, and the modelProvider→ModelProvider rename affects other callers. PR cannot be cleanly reverted; finish cleanup or provide a rollback plan.

**Comment by liliu-z** [comment]
**internal/proxy/rerank/operator_filter.go:1** FilterOp.Execute assumes the FunctionExpr chunked output shares chunk boundaries with the input DataFrame. Arrow does not guarantee this. Add an assertion that input.chunkSizes[chunkIdx]==boolChunk.Len() to fail fast.

**Comment by liliu-z** [comment]
**internal/proxy/rerank/converter.go:1** Missing unit tests: (1) empty path + offset/limit on chain pipeline, (2) end-to-end hybrid + chain-path strictGroupSize.

**Comment by liliu-z** [comment]
**internal/proxy/search_pipeline.go:1** Single-shard rerank score negation inverts sort order vs prior versions. Must be loudly flagged in release notes as a breaking behavior change.

**Comment by junjiejiangjjj** [comment]
Thanks for the additional pass. Quick rundown:

**1. `strictGroupSize` dropped** — Not a regression. Legacy `rerank.SearchParams` carries the field but no reranker ever reads it. Actual enforcement lives in segcore (`SearchGroupByOperator::IsGroupResEnough`); proxy/rerank just consumes what segcore returns. Both old and new paths are consistent.

**2. `ids==nil && totalRows>0`**
**3. `scores==nil && totalRows>0`** — Both unreachable. All reduce paths initialize `Ids` via `setupIdListForSearchResult` and append `Scores`/`Topks` 1:1 with `len(scores) == sum(topks) == totalRows` as an invariant. The `totalRows == 0` branches handle the only legitimate empty case. Happy to add fail-fast assertions if you want defense in depth, but no real path hits these.

**4. Legacy `rerank/` cleanup** — The remaining 12 files are the rerank shared infrastructure intentionally retained: `function_score.go` (`GetRerankName`/`BoostName` registry), `model_function.go` (`ModelProvider` interface + factory), and 8 concrete provider implementations — all actively used by `chain/rerank_builder.go` and `chain/expr/rerank_model_expr.go`. The `modelProvider` → `ModelProvider` rename doesn't appear in the diff; only the capitalized form exists on master and this branch.

**5. `FilterOp` chunk-boundary assertion** — Real defensive gap, but all current `FunctionExpr` implementations produce one bool chunk per input chunk so the assumption holds. Same shape as the `hybridAssemble` nil-computer branch we converted to fail-loud — happy to add the assertion if you want.

**6. Missing tests** — Will add (1) empty input + offset/limit on the chain pipeline. (2) follows from #1, so n/a.

**7. Single-shard score-negation behavior change** — Agreed. Will add a "Behavior change" entry to the PR description / release notes covering hybrid + group_by + distance metric (L2/HAMMING/JACCARD) + single-shard collections.

Let me know if any of #1/#2/#3/#5 are blocking — happy to add the assertions as pure defense in depth if preferred.
